src/Futhark/CodeGen/Backends/CCUDA.hs

```diff
@@ -337,12 +363,39 @@ callKernel (LaunchKernel safety kernel_name args num_blocks block_size) = do
         $id:time_start = get_wall_time();
       }
       $items:bef
-      CUDA_SUCCEED_OR_RETURN(
-        cuLaunchKernel(ctx->$id:kernel_name,
-                       grid[0], grid[1], grid[2],
-                       $exp:block_x, $exp:block_y, $exp:block_z,
-                       $exp:shared_tot, NULL,
-                       $id:args_arr, NULL));
+      if(ctx->use_multi_device){
+        for(int devID = 0; devID < ctx->cuda.device_count; devID++){
+            int localIdx = devID;
+            void *$id:args_arr[] = { $inits:args'' , &localIdx};
+            size_t grid_MD[3];
+            grid_MD[perm[0]] = $exp:grid_x_MD;
+            grid_MD[perm[1]] = $exp:grid_y_MD;
+            grid_MD[perm[2]] = $exp:grid_z_MD;
+            CUDA_SUCCEED_FATAL(cuCtxPushCurrent(ctx->cuda.contexts[devID]));
+            CUDA_SUCCEED_FATAL(cuLaunchKernel(ctx->$id:(kernelMultiDevice kernel_name)[devID],
+                                                  grid_MD[0], grid_MD[1], grid_MD[2],
+                                                  $exp:block_x, $exp:block_y, $exp:block_z,
+                                                  $exp:shared_tot, NULL,
+                                                  $id:args_arr, NULL));
+            CUDA_SUCCEED_FATAL(cuEventRecord(ctx->cuda.kernel_done[devID * 2
+                                             + ctx->cuda.kernel_iterator], NULL));
+            for(int other_dev = 0; other_dev < ctx->cuda.device_count; other_dev++){
```

Comment:
I am not at all sure what is going on here, and I remember we talked about it.  Any other reader would be completely mystified.

But more significantly, I'm not completely sure it is correct.  You allocate two event objects per device (right?).  How can we be sure only one kernel is in flight at a time?  Or doesn't that matter?  Is it OK that the host-side events are overwritten?

Reply:
The events exists to provide a kernel to kernel synchronization. If you have 2 kernels and put them on the same stream, you know that the first kernel is finished by the time kernel 2 starts and any data dependencies on kernel 2 from kernel 1 is fulfilled. If you launch kernels on multiple streams, which I'm forced to because I've multiple devices, I lose such assurances.
In our case is "divide" the kernels into smaller version. Say for example: kernel_1_1, kernel_1_2, kernel_1_2  kernel_2_2.
There might exists data dependencies from kernel_1_1 to kernel_2_2, hence kernel_1_1 needs to be finished before kernel_2_2 can be launched.

Hence the streams I create looks like this:

Lingo: Sx: Null stream on Device x, Kxy: Kernel_x_y, RDxEy: Record event y from device x, WDxEy: stream wait on device x event y

S1: K11 RD1E1 WD2E1 K21 RD1E2 WD2E2 K32 RD1E1 WD2E1 ...
S2: K12 RD2E1 WD1E1 K22 RD2E1 WD1E2 K32 RD2E1 WD1E1 ...

The re usage of events is ok, because re-recording events just overwrites the state of the event. which is also why we need two events because otherwise a stream could end up waiting on the same event twice and that would cause a deadlock.

The documentation for reusage of events: https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EVENT.html#group__CUDA__EVENT_1g95424d3be52c4eb95d83861b70fb89d1